# Hacker News Project


**Introduction**

Hacker News is a website started by incubator Y Combinator where users can submit stories that are voted and commented on. Hacker News is very popular within the technology and startup circle. Post that make the listing get hundreds of thousandd vistors as a result.

This project will explore the data sets of Show HN and Ask HN within the Hacker News website. Show HN are posts to show the Hacker News community a project, product or simply something interesting. Ask HN posts to ask the community a specific question. 

In [1]:
from csv import reader 

open_file = open('hacker_news.csv')
read_file = reader(open_file)
hn = list(read_file)

hn_header = hn[0]
hn_data = hn[1:]

print(hn_header)
print("\n")
print(hn_data[:5]) # First 5 rows of data for a visual 


['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


In [2]:
ask_posts = []
show_posts = []
other_posts = []

for post in hn:
    title = post[1]
    if title.lower().startswith('ask hn'):
        ask_posts.append(post)
    elif title.lower().startswith('show hn'):
        show_posts.append(post)
    else:
        other_posts.append(post)

print("These are the total number of ask hn posts: ", len(ask_posts))
print("\n")
print("These are the total number of show hn posts: ", len(show_posts))
print("\n")
print("These are the total number of other hn posts: ", len(other_posts))


These are the total number of ask hn posts:  1744


These are the total number of show hn posts:  1162


These are the total number of other hn posts:  17195


In [3]:
print(hn_header)
print('\n')
print("First five rows of ask hn posts from the list:")
print("\n")
print(ask_posts[:5])
print("\n")
print("First five rows of show hn posts from the list")
print("\n")
print(show_posts[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


First five rows of ask hn posts from the list:


[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14'], ['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20'], ['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']]


First five rows of show hn posts from the list


[['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03'], ['10646440', 'Show HN: Something pointless I made', 'htt

In [4]:
total_ask_comments = 0

for row in ask_posts:
    ask_comments = int(row[4])
    total_ask_comments += ask_comments

avg_ask_comments = total_ask_comments / len(ask_posts)

print("This is the average number of comments per ask hn post:", avg_ask_comments)

    

This is the average number of comments per ask hn post: 14.038417431192661


In [5]:
total_show_comments = 0

for row in show_posts:
    show_comments = int(row[4])
    total_show_comments += show_comments
    
avg_show_comments = total_show_comments / len(show_posts)

print("This is the average number of comments per show hn posts:", avg_show_comments)


This is the average number of comments per show hn posts: 10.31669535283993


We can see that ask posts receive more comments than show posts by 26.5 per cent. 

In [12]:
import datetime as dt 

result_list = []

for post in ask_posts:
    result_list.append([post[6], int(post[4])])
    
counts_by_hour = {}
comments_by_hour = {}
date_format = "%m/%d/%Y %H:%M" #This is the date format in the data set

for row in result_list:
    date = row[0] #Extracted from the dictionaries position
    comment = row[1] # ""
    #strftime at the end allows you to just show the time aka parse
    time = dt.datetime.strptime(date, date_format).strftime("%H")
    if time in counts_by_hour:
        comments_by_hour[time] += comment
        counts_by_hour[time] += 1 
    else:
        comments_by_hour[time] = comment
        counts_by_hour[time] = 1 
                       
print(comments_by_hour) 
print("\n")
print(counts_by_hour)

{'09': 251, '13': 1253, '10': 793, '14': 1416, '16': 1814, '23': 543, '12': 687, '17': 1146, '15': 4477, '21': 1745, '20': 1722, '02': 1381, '18': 1439, '03': 421, '05': 464, '19': 1188, '01': 683, '22': 479, '08': 492, '04': 337, '00': 447, '06': 397, '07': 267, '11': 641}


{'09': 45, '13': 85, '10': 59, '14': 107, '16': 108, '23': 68, '12': 73, '17': 100, '15': 116, '21': 109, '20': 80, '02': 58, '18': 109, '03': 54, '05': 46, '19': 110, '01': 60, '22': 71, '08': 48, '04': 47, '00': 55, '06': 44, '07': 34, '11': 58}


In [15]:
avg_by_hour = []

for count in counts_by_hour:
    avg_by_hour.append([count, comments_by_hour[count] / counts_by_hour[count]])

avg_by_hour

[['09', 5.5777777777777775],
 ['13', 14.741176470588234],
 ['10', 13.440677966101696],
 ['14', 13.233644859813085],
 ['16', 16.796296296296298],
 ['23', 7.985294117647059],
 ['12', 9.41095890410959],
 ['17', 11.46],
 ['15', 38.5948275862069],
 ['21', 16.009174311926607],
 ['20', 21.525],
 ['02', 23.810344827586206],
 ['18', 13.20183486238532],
 ['03', 7.796296296296297],
 ['05', 10.08695652173913],
 ['19', 10.8],
 ['01', 11.383333333333333],
 ['22', 6.746478873239437],
 ['08', 10.25],
 ['04', 7.170212765957447],
 ['00', 8.127272727272727],
 ['06', 9.022727272727273],
 ['07', 7.852941176470588],
 ['11', 11.051724137931034]]

In [25]:
# Create new list to display the average comments by hour to be positioned first
# in the list
swap_avg_by_hour = []

for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])

print(swap_avg_by_hour)

sorted_swap = sorted(swap_avg_by_hour, reverse=True)

sorted_swap

[[5.5777777777777775, '09'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [16.796296296296298, '16'], [7.985294117647059, '23'], [9.41095890410959, '12'], [11.46, '17'], [38.5948275862069, '15'], [16.009174311926607, '21'], [21.525, '20'], [23.810344827586206, '02'], [13.20183486238532, '18'], [7.796296296296297, '03'], [10.08695652173913, '05'], [10.8, '19'], [11.383333333333333, '01'], [6.746478873239437, '22'], [10.25, '08'], [7.170212765957447, '04'], [8.127272727272727, '00'], [9.022727272727273, '06'], [7.852941176470588, '07'], [11.051724137931034, '11']]


[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

In [27]:
#Sorting the values from top 5 hours
print('Top 5 Hours for Ask Posts Comments')

#To look through each average and hour a comma is needed in the for loop
for avg, hr in sorted_swap[:5]:
    print(
        "{}: {:.2f} average comments per post".format(
            dt.datetime.strptime(hr, "%H").strftime("%H:%M"),avg))
#strftime is to specify the format you want to use
#{:.2f} is used to format the average

Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


The hour that receives the most comments per ask post is 15:00 (3:00pm EST) with 38.59 comments per post with second place receiving 23.81 average comments per post at 2:00. There is about a 61% increase in the number of comments between the hours with the highest and second highest average number of comments. 

# Conclusion

In this project we analyzed ask posts and show posts to determine which of the two posts will receive more comments on average. Our analysis states that ask posts receive more comments than show posts. We would recommend posting around 3:00pm Eastern Time to receive the highest number of comments. 